### PR TITLE
Demotion modchip rework

### LINKED ISSUE #34213: demotion modchips suck ass
(WEB REPORT BY: janez420ivan REMOTE: 206.221.180.138:7777)
# Revision
9106c7e23d9d71f5e453fba894e4a1bb85521a5a

if an ID gets demoted via modchip, you can just re-add access in an ID machine

but you can't use another modchip to remove access again as it's already considered demoted(while it still has access)

### PR DESCRIPTION
**Pull request by SECBATON-GRIFFON**
[tweak]

## What this does
addresses problems in and closes #34213, by making demotion modchips a visible thing on ID cards that can be removed with the associated stamp, also cleans up the whole system for that. while on an ID, blocks changing access to these from identification computers.

## Why it's good
gives them more purpose.

## Changelog
:cl:
 * rscadd: Demotion modchips now block job, skin and access changes to their specified departments on identification computers.
 * rscadd: Demotion modchips now show up on IDs when examined and can be removed by stamping the ID with a matching stamp from them. Captain and HoP stamps will remove them regardless.
 * tweak: Name entry on demotion modchips has now been replaced with DNA entry or scanning it on the target ID.

### CODE REVIEW

**Comment by pedr69** [comment]
not sure if you've touched the card appearances yet
currently the `centcom_old` and `gold` colors are undemotable

**Comment by Eneocho** [comment]
> not sure if you've touched the card appearances yet currently the `centcom_old` and `gold` colors are undemotable

Well, both of them are only normally available on captain access-tier cards, which you shouldn't be able to demote anyways.

**Comment by SECBATON-GRIFFON** [comment]
should i remove that whole "set name of card to demote" thing before applying from the chips? it feels like an unnecessary extra step, could replace it with copying card biometrics.

**Comment by Eneocho** [comment]
Yeah the whole "set name of card" thing feels like a half baked feature. Copying the biometrics would work. 
There would be little issue with cards without biometrics, as those would require either a HoP or someone with captain level access to create, both of which can just demote the card (or just stun the guy, plop them into a scanner and use the card to set biometrics).

File: code/modules/paperwork/human_resources.dm [hunk 1]
```diff
@@ -13,136 +22,99 @@
...
 
 /obj/item/demote_chip/attack_self(mob/user as mob)
-	if(target_name != null) //Used hand-labeler as example
-		to_chat(user, "<span class='notice'>The target name cannot be reset!</span>")
+	if(target_dna) //Used hand-labeler as example
+		to_chat(user, "<span class='notice'>The target DNA cannot be reset!</span>")
 		return
 	else
-		var/str = copytext(reject_bad_text(input(user,"Enter the properly capitalized name for demotion","Set name","") as text|null),1,MAX_NAME_LEN)
+		var/str = sanitize(input(user,"Enter the DNA hash for demotion","Set DNA","") as text|null)
```
[the remark this answers is not in the document]
Comment (SECBATON-GRIFFON) [reply code/modules/paperwork/human_resources.dm:39]: i mean you wouldn't really, but someone would probably complain if i took it out altogether. if anything this is probably for cases where you don't have the ID on hand but do have medical records.

[the remark this answers is not in the document]
Comment (pedr69) [reply code/modules/paperwork/human_resources.dm:39]: i prefer the old name method
but the DNA thing lets traitor lawyers do some shenanigans so i guess it works too

**Pull request merged**